src/main/java/edu/kit/compiler/optimizations/attributes/AttributeAnalysis.java

```diff
@@ -0,0 +1,407 @@
+        if (cachedAttributes == null) {
+            var graph = entity.getGraph();
+            if (graph != null) {
+                // insert a set of dummy attributes for the function
+                var dummyAttrs = new Attributes(Purity.CONST, false, false);
+                functions.put(entity, dummyAttrs);
+
+                // First analyze the memory chains to determine purity
+                var attributes = new Attributes(Purity.CONST, true, false);
+                new MemoryVisitor(graph, attributes).apply();
+
+                // Second analyze termination behavior of the function
+                checkTermination(graph, attributes);
+
+                // Third analyze whether the function is malloc-like
+                attributes.setMalloc(isMallocLike(graph));
+
+                functions.put(entity, attributes);
+                return attributes;
+            } else if (entity.equals(calloc)) {
+                // special case for calls to calloc
+                var attributes = new Attributes(Purity.PURE, true, true);
+                functions.put(entity, attributes);
+                return attributes;
+            } else {
+                // no implementation known, e.g std library function
+                functions.put(entity, Attributes.MINIMUM);
+                return Attributes.MINIMUM;
+            }
+        } else {
+            return cachedAttributes;
+        }
+    }
+
+    /**
+     * Check the given graph for control flow loops and unset `terminates` in
+     * the given attributes if a loop is found.
+     */
+    private static void checkTermination(Graph graph, Attributes attributes) {
+        graph.incBlockVisited();
+        checkTermination(graph.getEndBlock(), attributes);
+
+        for (var keepAlive : graph.getEnd().getPreds()) {
+            if (keepAlive.getMode().equals(Mode.getBB())
+                    && keepAlive.getOpCode() != ir_opcode.iro_Bad) {
+                checkTermination((Block) keepAlive, attributes);
+            }
+        }
+    }
+
+    private static void checkTermination(Block initialBlock, Attributes attributes) {
+        var worklist = new StackWorklist<Block>(true);
+        worklist.enqueueInOrder(initialBlock);
+
+        while (attributes.isTerminates() && !worklist.isEmpty()) {
+            var block = worklist.dequeue();
+
+            if (block.blockVisited()) {
+                attributes.setTerminates(false);
+            } else {
+                block.markBlockVisited();
+                block.getPreds().forEach(pred -> {
+                    if (pred.getOpCode() != ir_opcode.iro_Bad) {
+                        worklist.enqueueInOrder((Block) pred.getBlock());
+                    }
+                });
+            }
+        }
+    }
+
+    /**
+     * Returns true if the given graph is malloc-like, i.e. it is guaranteed to
+     * always return newly allocated memory which is also alias free.
+     */
+    private boolean isMallocLike(Graph graph) {
```

Comment:
My understanding might be incorrect, but to me something seems to be quite wrong with this analysis.

The `malloc` attribute is used in `Attributes` to avoid deduplicating functions that return newly allocated memory. This is, as we are doing a conservative analysis, if there is any single case where the function returns newly allocated memory, we can not deduplicate it. However, the code here is doing quite the opposite: It checks whether it can proof that the function always returns new memory and _only_ in this case it forbids deduplication. So, I think this approximation is done from the wrong direction.

Or, rather, the current analysis could be useful for some (simple) alias analysis. But it is the wrong analysis for deciding whether a function can be deduplicated.

Reply:
You're absolutely right, that is not the correct analysis.

Reply:
I have changed the analysis accordingly. I have also had to remove the special handling for stores to newly allocated memory, as it does not work with the changed (or rather the correct) semantics of the `malloc` attribute.